hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/TestCapacitySchedulerNewQueueAutoCreationMixedMode.java

```diff
@@ -0,0 +1,226 @@
+    RMNodeLabelsManager mgr = new NullRMNodeLabelsManager();
+    mgr.init(csConf);
+    mockRM = new MockRM(csConf) {
+      protected RMNodeLabelsManager createNodeLabelManager() {
+        return mgr;
+      }
+    };
+    cs = (CapacityScheduler) mockRM.getResourceScheduler();
+    cs.updatePlacementRules();
+    mockRM.start();
+    cs.start();
+    autoQueueHandler = cs.getCapacitySchedulerQueueManager();
+    mockRM.registerNode("h1:1234", 32000, 32); // label = x
+  }
+
+  private void createMixedConfig() {
+    // root
+    // - a [memory=10%, vcores=1]
+    //   - a1 [memory=100%, vcores=100%]
+    //   - a2-auto [memory=2048, vcores=2]
+    // - b [memory=1024, vcores=10%]
+    // - c-auto [memory=1w, vcores=1w]
+    // - d [memory=20%, vcores=20%]
+    //   - d1-auto [memory=10%, vcores=1]
+    // - e-auto [memory=1w, vcores=10]
+    //   - e1-auto [memory=2048, vcores=2]
+
+    // Set up static queues
+    csConf.setQueues(CapacitySchedulerConfiguration.ROOT, new String[]{"a", "b", "d"});
+    csConf.setCapacityVector(A, NL, "[memory=10%, vcores=2]");
+    csConf.setCapacityVector(B, NL, "[memory=2000, vcores=10%]");
+    csConf.setCapacityVector(D, NL, "[memory=10%, vcores=10%]");
+    csConf.setQueues(A, new String[]{"a1"});
+    csConf.setCapacityVector(A1, NL, "[memory=100%, vcores=100%]");
+    csConf.setAutoQueueCreationV2Enabled(CapacitySchedulerConfiguration.ROOT, true);
+    csConf.setAutoQueueCreationV2Enabled(A, true);
+    csConf.setAutoQueueCreationV2Enabled(D, true);
+
+    // Set up dynamic queue templates
+    csConf.set(getTemplateKey(CapacitySchedulerConfiguration.ROOT, "capacity"), "[memory=2w, vcores=5w]");
+    csConf.set(getParentTemplateKey(CapacitySchedulerConfiguration.ROOT, "capacity"), "[memory=2w, vcores=10]");
+    csConf.set(getLeafTemplateKey(CapacitySchedulerConfiguration.ROOT + ".*", "capacity"), "[memory=2000, vcores=2]");
+    csConf.set(getLeafTemplateKey(D, "capacity"), "[memory=1000, vcores=1]");
+
+  }
+  /*
+   Create and validate the following structure with mixed resource vectors and non-legacy queue mode:
+
+                           root
+      ┌─────┬────────┬─────┴─────┬─────────┐
+      a     b      c-auto     e-auto       d
+      |                        |           |
+     a1                      e1-auto    d1-auto
+    */
+  @Test
+  public void testAutoCreateQueueRefresh() throws Exception {
+    createMixedConfig();
+    startScheduler();
+
+    createQueue("root.c-auto");
+
+    // Check if queue c-auto got created
+    CSQueue cAuto = cs.getQueue("root.c-auto");
+    // At this point queues a, b, d exists, ant c-autp was just created
+    // b takes 2000 MB from the cluster, a and d take up 10 + 10 = 20% (6000 MB, 6 vcore), so c-auto should get the rest
+    // (24000 MB, 24 vcore) because it's the only one with configured weights
+    Assert.assertEquals(24 / 32f, cAuto.getAbsoluteCapacity(), 1e-4);
+    Assert.assertEquals(-1f, cAuto.getQueueCapacities().getWeight(), 1e-6);
```

Comment:
an existing epsilon could be used here (or extract this to a variable)

Reply:
Done.